src/net/udp.rs

```diff
@@ -1,46 +1,63 @@
-use std::io;
-use std::net::{self, ToSocketAddrs};
+use std::io::{self, Cursor, ErrorKind, Error, Read, Write};
+use std::net::{self, SocketAddr, ToSocketAddrs};
 
-use super::{Packet, RawPacket, SocketState};
-use bincode::deserialize;
-
-use error::Result;
-
-const BUFFER_SIZE: usize = 1024;
+use super::{SocketState, NetworkConfig};
+use packet::{header, Packet, PacketProcessor};
+use error::{NetworkError, Result};
 
 pub struct UdpSocket {
     socket: net::UdpSocket,
     state: SocketState,
-    recv_buffer: [u8; BUFFER_SIZE],
+    recv_buffer: [u8;1452],
+    config: NetworkConfig,
+    packet_processor: PacketProcessor
 }
 
 impl UdpSocket {
     pub fn bind<A: ToSocketAddrs>(addr: A) -> io::Result<Self> {
         let socket = net::UdpSocket::bind(addr)?;
         let state = SocketState::new();
+        // TODO: add functionality to get config from file.
+        let config = NetworkConfig::default();
 
         Ok(UdpSocket {
             socket,
             state,
-            recv_buffer: [0; BUFFER_SIZE],
+            recv_buffer: [0;1452],
+            packet_processor: PacketProcessor::new(config.clone()),
+            config: config,
         })
     }
 
     pub fn recv(&mut self) -> io::Result<Option<Packet>> {
-        let (len, _addr) = self.socket.recv_from(&mut self.recv_buffer)?;
+        let (len, _addr) = { self.socket.recv_from(&mut self.recv_buffer)? };
+
         if len > 0 {
-            // TODO: Remove unwrap and funnel result error types
-            let raw_packet: RawPacket = deserialize(&self.recv_buffer[..len]).unwrap();
-            let packet = self.state.process_received(_addr, &raw_packet).unwrap();
-            Ok(Some(packet))
-        } else {
-            Ok(None)
+            let packet = self.recv_buffer[..len].to_owned();
+
+            self.packet_processor.process_data(packet, _addr, &mut self.state)
+        }else {
+            return Err (Error::new(ErrorKind::Other, "Received bytes where of length 0"));
         }
     }
 
-    pub fn send(&mut self, packet: Packet) -> Result<io::Result<usize>> {
-        let (addr, payload) = self.state.pre_process_packet(packet)?;
-        Ok(self.socket.send_to(&payload, addr))
+    pub fn send(&mut self, mut packet: Packet) -> Result<usize> {
+        let (addr, mut packet_data) = self.state.pre_process_packet(packet, &self.config)?;
+
+        let mut bytes_send: usize = 0;
```

Comment:
should be `bytes_sent` 😄 . Also why the `usize`?

Reply:
Yea maybe better to let the combiler desire no specific reason for that.